### DISCUSSION TITLE
Global using aliases

### DISCUSSION THREAD

**Discussion by nesteruk** [dotnet/csharplang#5925] [Language Ideas]
When working with numerics in C#, it is tedious to write `Sin()` (from `System.Math`) instead of using `sin()` (lowercase). It's possible to create an alias, but this alias is relevant to the current file only.

It would be good to be able to write something like

```
global using static sin = System.Math.Sin;
```

And have this propagate throughout the project.

**Comment by HaloFour** [comment]
The ability to alias members sounds interesting, but I am having trouble with the idea that `Sin` is more tedious than `sin`.

**Reply by nesteruk** [reply]
It's a particular pain point that mathematicians experience, because in mathematics, function names are lowercase — F# did it right, C# did not. But this is only an illustration. Consider something like

```
using static println = System.Console.WriteLine;
```

Same idea.

**Reply by HaloFour** [reply]
There is no "right" or "wrong", there are subjective stylistic preferences.  C# isn't math or F#, and agreed upon naming guidelines for both the [language](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#naming-conventions) and the [runtime](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions) are that accessible member names are in PascalCase.

An aliasing feature might have other utility, such as resolving ambiguities or making longer names much shorter, though, which would have more utility and could be compared to type aliasing which is possible today.

**Reply by nesteruk** [reply]
They are stylistic preferences only if you confine yourself to the world of C#. As soon as we're talking about math, there are conventions: nobody writes `Sin(x)` with a capital in any country on Earth, you wouldn't find a single scientific paper that capitalizes it.

**Reply by HaloFour** [reply]
The inverse argument is also true, and that's my point.  C# isn't math, nor does it try to adhere to its conventions.  It's not expected that you can write C# in the same manner that you write mathematical notation.  You'd need a _lot_ more than just function name aliases to achieve that.  If you were writing a paper on C# it's expected that you'd follow C#'s naming conventions, otherwise you'd likely just confuse the majority of the readers.

**Reply by tannergooding** [reply]
Likewise computer programming in general isn't math. You have to deal with rounding error, finite precision, and a number of other oddities which mean "normal" mathematical rules don't apply or only partially apply.

**Comment by PathogenDavid** [comment]
You can do this today if you just define your own class with the aliases you want and `global using static` it instead: ([SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCecAmIA1AD4DmANgPbAQUAEAAgAxMCMAbPQLI7cQIAFgG4AsAChGAJjYSJjAMzsu0nnwGCJAbwn09TJY0718VAK7AKqegGdYAClMWr9AB4BKXfu8BeAHw8GgB0AMoOHmLiAL5ykoZS9AAqqDYI2l56iibmltYAEqgU1PQ+tg5sQczukTHiQA==))

```csharp
global using static MyMath;
using System;

public static class MyMath
{
    public static double sin(double x)
        => Math.Sin(x);
}

public class Test
{
    public double Hello = sin(1.0);
}
```

**Reply by lonix1** [reply]
Thanks... I swiped your code for use on [SO](https://stackoverflow.com/a/74278742/9971404).
